src/tr1/game/output.c

```diff
@@ -93,6 +94,64 @@ static void M_CalcRoomVertices(const ROOM_MESH *mesh);
 static void M_CalcRoomVerticesWibble(const ROOM_MESH *mesh);
 static void M_CalcWibbleTable(void);
 
+static void M_DrawSphere(const XYZ_32 pos, const int32_t radius)
+{
+    GFX_Context_SetWireframeMode(true);
+
+    const RGBA_8888 color = { .r = 255, .g = 255, .b = 255, .a = 128 };
+
+    S_Output_DisableTextureMode();
+    S_Output_DisableDepthWrites();
+    S_Output_SetBlendingMode(GFX_BLEND_MODE_NORMAL);
+
+    // More subdivisions means smoother spheres.
+    const int32_t subdivisions = 12;
+    PHD_VBUF vertices[(subdivisions + 1) * (subdivisions + 1)];
+    int index = 0;
+
+    for (int i = 0; i <= subdivisions; i++) {
+        const float theta = (M_PI * i) / subdivisions; // Latitude angle
+        const float sin_theta = sinf(theta);
+        const float cos_theta = cosf(theta);
+
+        for (int j = 0; j <= subdivisions; j++) {
+            const float phi = (2 * M_PI * j) / subdivisions; // Longitude angle
+            const float sin_phi = sinf(phi);
+            const float cos_phi = cosf(phi);
+
+            // Convert spherical coordinates to 3D points.
+            XYZ_16 vertex_pos = {
+                .x = pos.x + radius * cos_phi * sin_theta,
+                .y = pos.y + radius * cos_theta,
+                .z = pos.z + radius * sin_phi * sin_theta,
+            };
+
+            M_CalcVertex(&vertices[index], vertex_pos);
+            vertices[index].g = HIGH_LIGHT;
+            index++;
+        }
+    }
+
+    for (int i = 0; i < subdivisions; i++) {
+        for (int j = 0; j < subdivisions; j++) {
+            const int index_0 = i * (subdivisions + 1) + j;
+            const int index_1 = (i + 1) * (subdivisions + 1) + j;
+            const int index_2 = (i + 1) * (subdivisions + 1) + (j + 1);
+            const int index_3 = i * (subdivisions + 1) + (j + 1);
+            S_Output_DrawFlatTriangle(
+                &vertices[index_0], &vertices[index_1], &vertices[index_2],
+                color);
+            S_Output_DrawFlatTriangle(
+                &vertices[index_0], &vertices[index_2], &vertices[index_3],
+                color);
+        }
+    }
+
+    S_Output_SetBlendingMode(GFX_BLEND_MODE_OFF);
+    S_Output_EnableDepthWrites();
+    GFX_Context_SetWireframeMode(false);
```

Comment:
This may cause issues when the user executes `/wireframe 1`, as it disables it for the rest of the pipeline. Even if this doesn't manifest right now, it's most likely due to the sequence of function calls. As such, I'd suggest changing this to restore the previous state.

Reply:
Good call. Did you see my comment earlier about this? Is there a better way to handle this?
`Applies wireframe mode during drawing which works well except for the fact that the color is ignored so the spheres are pretty opaque. I couldn't get using only glPolygonMode(GL_LINE) to work.`

Reply:
While I can't get the colors to show up as semitransparent no matter what I try in the `GL_LINE` mode, I noticed some issues with the depth buffer. A good area to test this is near the Scion pedestal in Qualopec. Removing the lines about toggling depth writes seems to help.